src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines.Networking.Tls.Internal.OpenSsl
+{
+    internal static unsafe class CustomBio
+    {
+        //Max Pipeline block size at the moment (4k - 1 cache line for tracking info) 
+        //this should come from the library directly when available in the future
+        const int MaxBlockSize = 1024 * 4 - 64;
+        const int BIO_TYPE_MEM = 1 | 0x0400 | 2;
+        const int BIO_TYPE_SOURCE_SINK = 0x0400;
+        static readonly bio_method_st _methodStruct;
+        static readonly IntPtr _methodPtr;
+        static readonly Create _create;
+        static readonly Write _write;
+        static readonly Read _read;
+        static readonly Control _control;
+        static readonly Free _free;
+
+        static CustomBio()
+        {
+            _create = CreateBio;
+            _write = WriteBio;
+            _read = ReadBio;
+            _control = ControlBio;
+            _free = FreeBio;
+
+            _methodStruct = new bio_method_st()
+            {
+                create = _create,
+                breadDelegate = _read,
+                bwriteDelegate = _write,
+                ctrlDelegate = _control,
+                type = BIO_TYPE_MEM,
+                destroy = _free,
+                name = null
+            };
+            var sizeToAlloc = Marshal.SizeOf(_methodStruct);
+            _methodPtr = Marshal.AllocHGlobal(sizeToAlloc);
+            Marshal.StructureToPtr(_methodStruct, _methodPtr, false);
+        }
+
+        public static IntPtr Custom() => _methodPtr;
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Create(ref bio_st bio);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Write(ref bio_st bio, void* buf, int num);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Read(ref bio_st bio, void* buf, int size);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate long Control(ref bio_st bio, BioControl cmd, long num, void* ptr);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Free(ref bio_st data);
+
+        private static class WindowsLib
+        {
+            public const string CryptoDll = "libeay32.dll";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static class UnixLib
+        {
+            public const string CryptoDll = "libcrypto.so";
```

Comment:
Unless we have scary binding magic, this will fail on any (package based distro) machine without -dev packages.  It needs to use the strong SONAME.  `libcrypto.so.1.0.0`.  And that'll fail on at least one distro, which doesn't follow OpenSSL's SONAME guidelines.  (This is part of what the shim protects against)